File: app/src/main/java/com/rowland/delivery/features/dash/presentation/adapters/ProductAdapter.java
package com.rowland.delivery.features.dash.presentation.adapters;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.chauthai.swipereveallayout.SwipeRevealLayout;
import com.chauthai.swipereveallayout.ViewBinderHelper;
import com.rowland.delivery.features.dash.domain.models.product.Product;
import com.rowland.delivery.features.dash.presentation.tools.recylerview.HFRecyclerView;
import com.rowland.delivery.merchant.R;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Rowland on 5/13/2018.
 */

public class ProductAdapter extends HFRecyclerView<Product> {

    ViewBinderHelper viewBinderHelper;

    public ProductAdapter() {
        this(false, false);
        viewBinderHelper = new ViewBinderHelper();
        viewBinderHelper.setOpenOnlyOne(true);
    }

    public ProductAdapter(boolean withHeader, boolean withFooter) {
        this(null, withHeader, withFooter);
    }

    public ProductAdapter(List<Product> data, boolean withHeader, boolean withFooter) {
        super(data, withHeader, withFooter);
    }

    @Override
    protected RecyclerView.ViewHolder getItemView(LayoutInflater inflater, ViewGroup parent) {
        ProductViewHolder x = new ProductViewHolder(inflater.inflate(R.layout.row_single_product, parent, false));
        return new ProductViewHolder(inflater.inflate(R.layout.row_single_product, parent, false));
    }

    @Override
    protected RecyclerView.ViewHolder getHeaderView(LayoutInflater inflater, ViewGroup parent) {
        return null;
    }

    @Override
    protected RecyclerView.ViewHolder getFooterView(LayoutInflater inflater, ViewGroup parent) {
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ProductViewHolder) {
            ProductViewHolder itemHolder = (ProductViewHolder) holder;
            Product data = getItem(position);
            itemHolder.bind(data);

        } else if (holder instanceof HeaderViewHolder) {

        } else if (holder instanceof FooterViewHolder) {

        }
    }

    public void setList(final List<Product> products) {
        if (this.data == null) {
            this.data = products;
            notifyItemRangeInserted(0, products.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return ProductAdapter.this.data.size();
                }

                @Override
                public int getNewListSize() {
                    return products.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return ProductAdapter.this.data.get(oldItemPosition).getFirestoreUid() ==
                            products.get(newItemPosition).getFirestoreUid();
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    Product newItem = products.get(newItemPosition);
                    Product oldItem = products.get(oldItemPosition);
                    return newItem.getFirestoreUid() == oldItem.getFirestoreUid()
                            && newItem.equals(oldItem);
                }
            });
            this.data = products;
            result.dispatchUpdatesTo(this);
        }
    }

    public void saveStates(Bundle outState) {
        viewBinderHelper.saveStates(outState);
    }

    public void restoreStates(Bundle inState) {
        viewBinderHelper.restoreStates(inState);
    }

    public class ProductViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.swipe_layout)
        SwipeRevealLayout swpRevealLayout;

        @BindView(R.id.product_imageview)
        ImageView productImage;

        @BindView(R.id.product_name_textview)
        TextView productName;

        @BindView(R.id.product_description)
        TextView productDescription;

        @BindView(R.id.product_price)
        TextView productPrice;

        @BindView(R.id.txt_unpublish)
        TextView txtUnpublish;

        @BindView(R.id.txt_edit)
        TextView txtEdit;

        public ProductViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void bind(Product product) {
            Glide.with(productImage.getContext()).load(product.getImageUrl()).centerCrop().crossFade().into(productImage);
            productName.setText(product.getName());
            productDescription.setText(product.getDescription());
            productPrice.setText("Ksh " + product.getPrice());

            viewBinderHelper.bind(swpRevealLayout, product.getFirestoreUid());
            txtUnpublish.setOnClickListener(v -> {
                //
                viewBinderHelper.closeLayout(product.getFirestoreUid());
            });

            txtEdit.setOnClickListener(v -> {
                //
                viewBinderHelper.closeLayout(product.getFirestoreUid());
            });
        }
    }

    public static class HeaderViewHolder extends RecyclerView.ViewHolder {
        public HeaderViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class FooterViewHolder extends RecyclerView.ViewHolder {
        public FooterViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/com/rowland/delivery/features/dash/presentation/adapters/ProductAdapter.kt
package com.rowland.delivery.features.dash.presentation.adapters

import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.chauthai.swipereveallayout.ViewBinderHelper
import com.github.florent37.glidepalette.BitmapPalette
import com.github.florent37.glidepalette.GlidePalette
import com.google.firebase.storage.FirebaseStorage
import com.rowland.delivery.features.dash.presentation.model.ProductModel
import com.rowland.delivery.features.dash.presentation.tools.recylerview.HFRecyclerView
import com.rowland.delivery.features.dash.presentation.viewmodels.product.ProductEvent
import com.rowland.delivery.merchant.R
import kotlinx.android.synthetic.main.content_single_product.view.*
import kotlinx.android.synthetic.main.content_single_product_reveal.view.*
import kotlinx.android.synthetic.main.row_single_product.view.*


/**
 * Created by Rowland on 5/13/2018.
 */

class ProductAdapter(data: List<ProductModel>?, withHeader: Boolean, withFooter: Boolean) : HFRecyclerView<ProductModel>(data, withHeader, withFooter) {

    private lateinit var viewBinderHelper: ViewBinderHelper
    private lateinit var actionListener: ProductActionListener

    interface ProductActionListener {
        fun onProductActionListener(event: ProductEvent, fn: () -> Unit)
    }

    constructor() : this(false, false) {
        viewBinderHelper = ViewBinderHelper()
        viewBinderHelper.setOpenOnlyOne(true)
    }

    constructor(withHeader: Boolean, withFooter: Boolean) : this(null, withHeader, withFooter)

    override fun getItemView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        return ProductViewHolder(inflater.inflate(R.layout.row_single_product, parent, false))
    }

    override fun getHeaderView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder? {
        return null
    }

    override fun getFooterView(inflater: LayoutInflater, parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder? {
        return null
    }

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        if (holder is ProductViewHolder) {
            val data = getItem(position)
            holder.bind(data)

        } else if (holder is HeaderViewHolder) {

        } else if (holder is FooterViewHolder) {

        }
    }

    fun setActionListener(actionListener: ProductActionListener) {
        this.actionListener = actionListener
    }

    fun setList(productEntities: List<ProductModel>) {
        if (this.data == null) {
            this.data = productEntities
            notifyItemRangeInserted(0, productEntities.size)
        } else {
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return this@ProductAdapter.data.size
                }

                override fun getNewListSize(): Int {
                    return productEntities.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return this@ProductAdapter.data[oldItemPosition].firestoreUid === productEntities[newItemPosition].firestoreUid
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val newItem = productEntities[newItemPosition]
                    val oldItem = productEntities[oldItemPosition]
                    return newItem.firestoreUid === oldItem.firestoreUid && newItem == oldItem
                }
            })
            this.data = productEntities
            result.dispatchUpdatesTo(this)
        }
    }

    fun saveStates(outState: Bundle) {
        viewBinderHelper.saveStates(outState)
    }

    fun restoreStates(inState: Bundle) {
        viewBinderHelper.restoreStates(inState)
    }

    inner class ProductViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {

        init {
            ButterKnife.bind(this, itemView)
        }

        fun bind(ProductModel: ProductModel) {
            FirebaseStorage.getInstance().reference.child(ProductModel.imageUrls.get(ProductModel.imageUrls.size -1)).downloadUrl
                    .addOnSuccessListener { uri ->
                        val options = RequestOptions()
                        options.centerCrop().diskCacheStrategy(DiskCacheStrategy.ALL)

                        Glide.with(itemView.product_imageview!!.context)
                                .load(uri.toString())
                                .apply(options)
                                .listener(GlidePalette.with(uri.toString())
                                        .use(BitmapPalette.Profile.VIBRANT_LIGHT)
                                        .intoBackground(itemView.card_main_content)
                                        .use(BitmapPalette.Profile.VIBRANT_DARK)
                                        .intoTextColor(itemView.product_description, BitmapPalette.Swatch.BODY_TEXT_COLOR)
                                        .crossfade(true))
                                .into(itemView.product_imageview!!)
                    }

            itemView.product_name_textview!!.text = ProductModel.name
            itemView.product_description!!.text = ProductModel.description
            itemView.product_price!!.text = "Ksh " + ProductModel.price!!

            viewBinderHelper.bind(itemView.swipe_layout!!, ProductModel.firestoreUid)

            itemView.btn_unpublish!!.setOnClickListener { v ->
                itemView.btn_unpublish!!.setIndeterminate()
                actionListener.onProductActionListener(ProductEvent.Unpublish(), { itemView.btn_unpublish.setFinish() })
                viewBinderHelper.closeLayout(ProductModel.firestoreUid)
            }

            itemView.btn_edit!!.setOnClickListener { v ->
                itemView.btn_edit!!.setIndeterminate()
                actionListener.onProductActionListener(ProductEvent.Edit(), { itemView.btn_edit.setFinish() })
                viewBinderHelper.closeLayout(ProductModel.firestoreUid)
            }
        }
    }

    class HeaderViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView)

    class FooterViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView)
}
